# Byte Ordering

In computing, byte order refers to the way in which bytes are arranged in a multi-byte data type, such as an integer or a float. There are two common byte orders: big-endian and little-endian. In big-endian byte order, the most significant byte (the byte with the highest value) comes first, while in little-endian byte order, the least significant byte (the byte with the lowest value) comes first.

## Finding my systems Byte Ordering

In [3]:
import sys

if sys.byteorder == 'little':
    print("Little-endian byte order")
else:
    print("Big-endian byte order")


Little-endian byte order


## Big Endian

In [1]:
import struct

# Create a 4-byte integer value in big-endian byte order
value = 0x12345678
packed_value = struct.pack('>I', value)

# Print the packed value in hexadecimal format
print(packed_value.hex())

# Unpack the packed value and print the integer value
unpacked_value = struct.unpack('>I', packed_value)
print(unpacked_value[0])


12345678
305419896


In the above example, we use the struct module in Python to pack a 4-byte integer value (0x12345678) into a byte string in big-endian byte order (>I). We then print the packed value in hexadecimal format (12345678). Finally, we use the struct module again to unpack the packed value back into an integer value in Python and print it (305419896). Note that the unpacked value is different from the original value because it is in the machine's native byte order (which is typically little-endian).

## Little Endian

In [2]:
import struct

# Create a 4-byte integer value in little-endian byte order
value = 0x12345678
packed_value = struct.pack('<I', value)

# Print the packed value in hexadecimal format
print(packed_value.hex())

# Unpack the packed value and print the integer value
unpacked_value = struct.unpack('<I', packed_value)
print(unpacked_value[0])


78563412
305419896


In the above example, we use the struct module again to pack the same 4-byte integer value (0x12345678) into a byte string, but this time in little-endian byte order (<I). We then print the packed value in hexadecimal format (78563412). Finally, we use the struct module again to unpack the packed value back into an integer value in Python and print it (305419896). Note that the unpacked value is the same as in the previous example, but the packed value is different because the byte order is different.